primer/src/Primer/TypeDef.hs

```diff
@@ -106,26 +105,21 @@ typeDefAST = \case
 typeDefKind :: TypeDef b () -> Kind' ()
 typeDefKind = foldr (KFun () . snd) (KType ()) . typeDefParameters
 
--- | A traversal over the contstructor fields in an typedef.
-_typedefFields :: Traversal (TypeDef b c) (TypeDef b' c) (Type' b ()) (Type' b' ())
-_typedefFields =
-  #_TypeDefAST
-    % #astTypeDefConstructors
-    % traversed
-    % #valConArgs
-    % traversed
-
-_typedefParamKinds :: Traversal (TypeDef b c) (TypeDef b c') (Kind' c) (Kind' c')
-_typedefParamKinds = traversalVL $ \f -> \case
-  TypeDefPrim td -> TypeDefPrim <$> traverseOf (#primTypeDefParameters % traversed % _2) f td
-  TypeDefAST td -> TypeDefAST <$> traverseOf (#astTypeDefParameters % traversed % _2) f td
-
 forgetTypeDefMetadata :: TypeDef b c -> TypeDef () ()
-forgetTypeDefMetadata =
-  over _typedefFields forgetTypeMetadata
-    . over _typedefParamKinds forgetKindMetadata
+forgetTypeDefMetadata (TypeDefPrim td) = TypeDefPrim $ over (#primTypeDefParameters % mapped % _2) forgetKindMetadata td
+forgetTypeDefMetadata (TypeDefAST (ASTTypeDef ps cs hs)) =
+  TypeDefAST $
+    ASTTypeDef
+      (over (mapped % _2) forgetKindMetadata ps)
+      (over (mapped % #valConArgs % mapped) forgetTypeMetadata cs)
+      hs
```

Comment:
It is unfortunate that this has got more verbose. I cannot see how to do this nice and lens-y (compare with old code above) because we now cannot change metadata in params independently to in ctors, as both controlled by same `c`. (The same comment applies to `generateTypeDefIDs`)

Reply:
Is it possible to at least write a lens which is as general as possible in updating kinds in all positions, and use that here?

Reply:
I did wonder about that, and decided that it would be a one-off slightly odd definition and not worth it. However, your comment made me realise that we can just use generic optics, which I think is actually pretty nice:
```haskell
forgetTypeDefMetadata :: TypeDef b c -> TypeDef () ()
forgetTypeDefMetadata =
  set (param @1) ()
    . set (param @0) ()

generateTypeDefIDs :: MonadFresh ID m => TypeDef () () -> m (TypeDef TypeMeta KindMeta)
generateTypeDefIDs =
  traverseOf (param @1) (\() -> meta)
    <=< traverseOf (param @0) (\() -> kmeta)
```

Reply:
Aha, yeah, of course. Nice!